### DISCUSSION TITLE
Ignore some blocks during spell checking

### DISCUSSION THREAD

**Discussion by BlueCutOfficial** [retextjs/retext#89] [Q&A] [answered]
Hello, 
I am looking for some advice about how to approach an issue with spell-checking. 

I work on a project that uses `remark-retext` with some `remark-lint` plugins and `retext-spell` to spell-check markdown files in my native language. My project contains links to English articles, and to avoid any confusion, I want to keep the `[title of the link](...)` in English to let my reader know the article is in English. 

It doesn't feel natural to me to add all the terms of the title ("title" "of" "the" "link") to the personal dictionary of the project because each word is not a proper exception, it's the whole title in one particular location that I want to ignore.

I would love to have a way to ignore a markdown block with something like this, [as documented in remark-lint](https://github.com/remarkjs/remark-lint#ignore-warnings):
```
<!-- lint ignore -->
[My English link](...)
```
but the above doesn't seem to apply to `retext` plugins.

Did I miss an easy way to ignore blocks with `retext-spell`? or what would be the best approach to handle this? 🙏

**Comment by wooorm** [comment, marked as answer]
Hi!

Hmm, using English labels in links, in otherwise, say, French, docs, seems a bit weird to me?
As an example, on a French Wikipedia page, you’ll see French text in links, even when they go to English-language pages.
If you’d want to notify readers, I’d personally probably add ``[xxx (pages en anglais)](yyy)`` or so?

If you’d want to ignore all links, you could use the [`source` option](https://github.com/syntax-tree/mdast-util-to-nlcst#source). `source: ['inlineCode', 'link', 'linkReference']`

To use comments to control messages, which is what `remark-lint` does for `lint` messages, you can use [`remark-message-control`](https://github.com/remarkjs/remark-message-control).
You can configure it to look for `<!--spell -->` comments with `name: 'spell'`, and then to control messages from `retext-spell` you’d set `sources: ['retext-spell']`!

**Reply by BlueCutOfficial** [reply]
Hi!
Thank you for the insight! 🙏 I am going to have a look at the solution you suggest

EDIT: we went for the last strategy, that perfectly matches our needs
> configure remark-message-control to look for  comments with name: 'spell', and then to control messages from retext-spell you’d set sources: ['retext-spell']!

**Reply by mansona** [reply]
So we ended up figuring this out (after a few pairing sessions 🤪). The problem we were facing was that we were getting our wires crossed between `unified-message-control` and  `remark-message-control`. We added `remark-message-control` to the end of the chain of unified with a `.use()` statement but really we needed to list `remark-message-control` as a normal remark plugin like this: 

```
["remark-message-control", { name: 'spell', source: ['retext-spell'] }],
```

once we did this then it all started working 🎉 I was also a bit suspicious about the use of `source` since you were suggesting things like `link` and `inlineCode` etc so it felt odd setting the source to be a different plugin 🤔 but it turns out that this whole thing just doesn't work without the source being set as `retext-spell` 🤷

**Reply by wooorm** [reply]
Oh, there are two `source` options. The one mentioned and linked above is for `toNlcst` / `remark-retext`, it’s a list of node types to treat as raw text, not normal prose.

The `sources` (note the `s`) option for `remark-message-control` and `unified-message-control` is different, it’s about message sources. Every tool emitting messages will annotate they package name, e.g., `my-package` as the source of the error.
`message-control` will look for those specifically.

As for where to place the plugins, that depends: `message-control` will look for nodes that configure things. And it needs to be able to find those nodes, those comments. They can be found in mdast/remark/markdown land, not in nlcst/retext/natural language!
